tool/tsh/azure.go

```diff
@@ -0,0 +1,423 @@
+			return trace.NewAggregate(err, cerr)
+		}
+		return trace.Wrap(err)
+	}
+
+	go func() {
+		if err := a.localALPNProxy.StartHTTPAccessProxy(a.cf.Context); err != nil {
+			log.WithError(err).Errorf("Failed to start local ALPN proxy.")
+		}
+	}()
+	return nil
+}
+
+// startLocalForwardProxy starts the local forward proxy.
+func (a *azureApp) startLocalForwardProxy(port string) error {
+	listenAddr := "localhost:0"
+	if port != "" {
+		listenAddr = fmt.Sprintf("localhost:%s", port)
+	}
+
+	// Note that the created forward proxy serves HTTP instead of HTTPS, to
+	// eliminate the need to install temporary CA for various Azure clients.
+	listener, err := net.Listen("tcp", listenAddr)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	a.localForwardProxy, err = alpnproxy.NewForwardProxy(alpnproxy.ForwardProxyConfig{
+		Listener:     listener,
+		CloseContext: a.cf.Context,
+		Handlers: []alpnproxy.ConnectRequestHandler{
+			// Forward Azure requests to ALPN proxy.
+			alpnproxy.NewForwardToHostHandler(alpnproxy.ForwardToHostHandlerConfig{
+				MatchFunc: alpnproxy.MatchAzureRequests,
+				Host:      a.localALPNProxy.GetAddr(),
+			}),
+
+			// Forward non-Azure requests to user's system proxy, if configured.
+			alpnproxy.NewForwardToSystemProxyHandler(alpnproxy.ForwardToSystemProxyHandlerConfig{
+				InsecureSystemProxy: a.cf.InsecureSkipVerify,
+			}),
+
+			// Forward non-Azure requests to their original hosts.
+			alpnproxy.NewForwardToOriginalHostHandler(),
+		},
+	})
+	if err != nil {
+		if cerr := listener.Close(); cerr != nil {
+			return trace.NewAggregate(err, cerr)
+		}
+		return trace.Wrap(err)
+	}
+
+	go func() {
+		if err := a.localForwardProxy.Start(); err != nil {
+			log.WithError(err).Errorf("Failed to start local forward proxy.")
+		}
+	}()
+	return nil
+}
+
+func printAzureIdentities(identities []string) {
+	if len(identities) == 0 {
+		return
+	}
+
+	sort.Strings(identities)
+
+	t := asciitable.MakeTable([]string{"Available Azure identities"}, identities)
+	fmt.Println(t.AsBuffer().String())
+}
+
+func getAzureIdentityFromFlags(cf *CLIConf, profile *client.ProfileStatus) (string, error) {
+	identities := profile.AzureIdentities
+	if len(identities) == 0 {
+		return "", trace.BadParameter("no Azure identities available, check your permissions")
+	}
+
+	// if flag is missing, try to find singleton identity; failing that, print available options.
+	if cf.AzureIdentity == "" {
+		if len(identities) == 1 {
+			log.Infof("Azure identity %v is selected by default as it is the only role configured for this azure app.", identities[0])
+			return identities[0], nil
+		}
+
+		printAzureIdentities(identities)
+		return "", trace.BadParameter("--azure-identity flag is required")
+	}
+
+	// exact match?
+	for _, identity := range identities {
+		if identity == cf.AzureIdentity {
+			return identity, nil
+		}
+	}
+
+	// suffix match?
+	expectedSuffix := fmt.Sprintf("/Microsoft.ManagedIdentity/userAssignedIdentities/%v", cf.AzureIdentity)
```

Comment:
If the user is copying the identity resource ID from the JSON view (the link on the right in the screenshot below), the resource ID will be lowercased (e.g., `/subscriptions/0000000/resourcegroups/paulgottschling/providers/microsoft.managedidentity/userassignedidentities/teleport-az`). Should this function do a case-insensitive comparison?

<img width="1254" alt="image" src="https://user-images.githubusercontent.com/25250668/204902412-9dc70d4a-2417-422a-a13e-e9b6ad8bbf15.png">

Reply:
@ptgott I've made changes both to `tsh app login` (so that "teleport-az" matches "/subscriptions/0000000/resourcegroups/paulgottschling/providers/microsoft.managedidentity/userassignedidentities/Teleport-AZ" ) and to RBAC engine (mainly to ensure that "Allow: foo" gets denied by "Deny: FOO").